astronomer/providers/amazon/aws/triggers/emr.py

```diff
@@ -0,0 +1,76 @@
+import asyncio
+from typing import Any, AsyncIterator, Dict, Iterable, Optional, Tuple
+
+from airflow.triggers.base import BaseTrigger, TriggerEvent
+
+from astronomer.providers.amazon.aws.hooks.emr import EmrStepSensorHookAsync
+
+
+class EmrStepSensorTrigger(BaseTrigger):
+    """
+    A trigger that fires once AWS Emr cluster step reach either target or failed state
+
+    :param job_flow_id: job_flow_id which contains the step check the state of
+    :param step_id: step to check the state of
+    :param aws_conn_id: aws connection to use, defaults to 'aws_default'
+    :param poke_interval: Time in seconds to wait between two consecutive call to
+        check emr cluster step state
+    :param target_states: the target states, sensor waits until
+        step reaches any of these states
+    :param failed_states: the failure states, sensor fails when
+        step reaches any of these states
+    """
+
+    def __init__(
+        self,
+        job_flow_id: str,
+        step_id: str,
+        aws_conn_id: str,
+        poke_interval: float,
+        target_states: Optional[Iterable[str]] = None,
+        failed_states: Optional[Iterable[str]] = None,
+    ):
+        super().__init__()
+        self.job_flow_id = job_flow_id
+        self.step_id = step_id
+        self.aws_conn_id = aws_conn_id
+        self.poke_interval = poke_interval
+        self.target_states = target_states or ["COMPLETED"]
+        self.failed_states = failed_states or ["CANCELLED", "FAILED", "INTERRUPTED"]
+
+    def serialize(self) -> Tuple[str, Dict[str, Any]]:
+        """Serializes EmrStepSensorTrigger arguments and classpath."""
+        return (
+            "astronomer.providers.amazon.aws.triggers.emr.EmrStepSensorTrigger",
+            {
+                "job_flow_id": self.job_flow_id,
+                "step_id": self.step_id,
+                "aws_conn_id": self.aws_conn_id,
+                "poke_interval": self.poke_interval,
+                "target_states": self.target_states,
+                "failed_states": self.failed_states,
+            },
+        )
+
+    async def run(self) -> AsyncIterator["TriggerEvent"]:  # type: ignore[override]
+        """Run until AWS Emr cluster step reach target or failed state"""
+        hook = EmrStepSensorHookAsync(
+            aws_conn_id=self.aws_conn_id, job_flow_id=self.job_flow_id, step_id=self.step_id
+        )
+        try:
+            while True:
```

Comment:
`while True` is not recommended usually. Is there a way we can have some breaking condition check here?

Reply:
Yes, this is an infinite loop. currently, docstring has some hints about the breaking condition. It will break if Either the EMR cluster reaches the target or failed state or AWS API throws an error or task timeout.